### PR TITLE
Write  ControllerRevisionLifecycleTest +7 Endpoints

### LINKED ISSUE #110121: Write  ControllerRevisionLifecycleTest +7 Endpoints
# Progress <code>[6/6]</code>

-   [X] APISnoop org-flow: [Appsv1ControllerRevisionLifecycleTest.org](https://github.com/apisnoop/ticket-writing/blob/master/Appsv1ControllerRevisionLifecycleTest.org)
-   [X] Test approval issue: #110121
-   [X] Test PR: #110122 
-   [x] Two weeks soak start date: [testgrid-link](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=5&graph-metrics=test-duration-minutes&include-filter-by-regex=should%20manage%20the%20lifecycle%20of%20a%20ControllerRevision) 8 Jul 2022
-   [x] Two weeks soak end date: 22 Jul 2022
-   [x] Test promotion PR: #111449

# Identifying an untested feature Using APISnoop

According to this APIsnoop query, there are still some remaining Controller Revision endpoints which are untested.

```sql-mode
    SELECT
      endpoint,
      path,
      kind
      FROM testing.untested_stable_endpoint
      where eligible is true
      and endpoint ilike '%ControllerRevision%'
      order by kind, endpoint desc
      limit 10;
```

```example
                        endpoint                      |                              path                               |        kind
  ----------------------------------------------------+-----------------------------------------------------------------+--------------------
   replaceAppsV1NamespacedControllerRevision          | /apis/apps/v1/namespaces/{namespace}/controllerrevisions/{name} | ControllerRevision
   readAppsV1NamespacedControllerRevision             | /apis/apps/v1/namespaces/{namespace}/controllerrevisions/{name} | ControllerRevision
   patchAppsV1NamespacedControllerRevision            | /apis/apps/v1/namespaces/{namespace}/controllerrevisions/{name} | ControllerRevision
   listAppsV1ControllerRevisionForAllNamespaces       | /apis/apps/v1/controllerrevisions                               | ControllerRevision
   deleteAppsV1NamespacedControllerRevision           | /apis/apps/v1/namespaces/{namespace}/controllerrevisions/{name} | ControllerRevision
   deleteAppsV1CollectionNamespacedControllerRevision | /apis/apps/v1/namespaces/{namespace}/controllerrevisions        | ControllerRevision
   createAppsV1NamespacedControllerRevision           | /apis/apps/v1/namespaces/{namespace}/controllerrevisions        | ControllerRevision
  (7 rows)

```

# API Reference and feature documentation

-   [Kubernetes API Reference Docs](https://kubernetes.io/docs/reference/kubernetes-api/)
-   [Kubernetes API > Workload Resources > Controller Revision](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/controller-revision-v1/)
-   [client-go: apps/v1/controllerrevision.go](https://github.com/kubernetes/client-go/blob/master/kubernetes/typed/apps/v1/controllerrevision.go#L42-L54)

# Test outline

```
Feature: Test create, delete, deleteCollection, list(All Namespaces), patch, read and replace ControllerRevision api endpoints
```

-   listAppsV1ControllerRevisionForAllNamespaces

```
Scenario: confirm that the list(All Namespaces) action will find a list of controllerRevisions
  Given the e2e test has as a running daemonset
  When the test lists all controller revisions by a label selector in all namespaces
  Then the requested action is accepted without any error
  And a list of ControllerRevisions must be returned
```

-   readAppsV1NamespacedControllerRevision

```
Scenario: confirm that the read action will find the details of a single controllerRevision
  Given the e2e test has a list of controllerRevisions
  When the test reads a controllerRevision from the list of ControllerRevisions
  Then the requested action is accepted without any error
  And the controllerRevision returned is not nil
```

-   patchAppsV1NamespacedControllerRevision

```
Scenario: confirm that the patch action will apply a change to a controllerRevision
  Given the e2e test has the current controllerRevision for the DaemonSet
  And a payload has been created with a new label
  When the test applies the patch to the controllerRevision
  Then the requested action is accepted without any error
  And the newly applied label is found
```

-   createAppsV1NamespacedControllerRevision

```
Scenario: confirm that the create action will add a new controllerRevision
  Given the e2e test has the patched controllerRevision for the DaemonSet
  When the test creates a new controllerRevision
  Then the requested action is accepted without any error
  And two controllerRevisions are found
```

-   deleteAppsV1NamespacedControllerRevision

```
Scenario: confirm that the delete action will remove a controllerRevision
  Given the e2e test has the two controllerRevisions for the DaemonSet
  When the test deletes the initial controllerRevision
  Then the requested action is accepted without any error
  And only one controllerRevision is found
```

-   replaceAppsV1NamespacedControllerRevision

```
Scenario: confirm that the replace action will apply the changes to a controllerRevision
  Given the e2e test has a single controllerRevision for the DaemonSet
  When the test updates the controllerRevision label
  Then the requested action is accepted without any error
  And change to the label is found in the controllerRevision
```

-   deleteAppsV1CollectionNamespacedControllerRevision

```
Scenario: confirm that deleteCollection action will remove a controllerRevision
  Given the e2e test has the updated controllerRevision for the DaemonSet
  And a new controllerRevision is created after patching the DaemonSet
  When the test applies the deleteCollection action with a labelSelector
  Then the requested action is accepted without any error
  And only one controllerRevision is found
```

# E2E Test

Using a number of existing e2e test practices a new [ginkgo test](https://github.com/ii/kubernetes/blob/create-controller-revision-test/test/e2e/apps/controller_revision.go#L109-L225) has been created for 7 ControllerRevision endpoints. The e2e logs for this test are listed below.

```
[It] should manage the lifecycle of a ControllerRevision
  /home/ii/go/src/k8s.io/kubernetes/test/e2e/apps/controller_revision.go:109
STEP: Creating DaemonSet "e2e-cf8wj-daemon-set"
STEP: Check that daemon pods launch on every node of the cluster.
May 19 10:10:19.766: INFO: Number of nodes with available pods controlled by daemonset e2e-cf8wj-daemon-set: 0
May 19 10:10:19.766: INFO: Node e2e-cr-control-plane-qkhlk is running 0 daemon pod, expected 1
May 19 10:10:20.773: INFO: Number of nodes with available pods controlled by daemonset e2e-cf8wj-daemon-set: 0
May 19 10:10:20.773: INFO: Node e2e-cr-control-plane-qkhlk is running 0 daemon pod, expected 1
May 19 10:10:21.773: INFO: Number of nodes with available pods controlled by daemonset e2e-cf8wj-daemon-set: 0
May 19 10:10:21.773: INFO: Node e2e-cr-control-plane-qkhlk is running 0 daemon pod, expected 1
May 19 10:10:22.773: INFO: Number of nodes with available pods controlled by daemonset e2e-cf8wj-daemon-set: 1
May 19 10:10:22.773: INFO: Number of running nodes: 1, number of available pods: 1 in daemonset e2e-cf8wj-daemon-set
STEP: Confirm DaemonSet "e2e-cf8wj-daemon-set" successfully created with "daemonset-name=e2e-cf8wj-daemon-set" label
STEP: Listing all ControllerRevisions with label "daemonset-name=e2e-cf8wj-daemon-set"
May 19 10:10:22.782: INFO: Located ControllerRevision: "e2e-cf8wj-daemon-set-78d45fff97"
STEP: Patching ControllerRevision "e2e-cf8wj-daemon-set-78d45fff97"
May 19 10:10:22.789: INFO: e2e-cf8wj-daemon-set-78d45fff97 has been patched
STEP: Create a new ControllerRevision
May 19 10:10:22.792: INFO: Created ControllerRevision: e2e-cf8wj-daemon-set-bb6fd6fcbSTEP: Confirm that there are two ControllerRevisions
May 19 10:10:22.792: INFO: Requesting list of ControllerRevisions to confirm quantityMay 19 10:10:22.794: INFO: Found 2 ControllerRevisions
STEP: Deleting ControllerRevision "e2e-cf8wj-daemon-set-78d45fff97"
STEP: Confirm that there is only one ControllerRevision
May 19 10:10:22.796: INFO: Requesting list of ControllerRevisions to confirm quantity
May 19 10:10:22.798: INFO: Found 1 ControllerRevisions
STEP: Updating ControllerRevision "e2e-cf8wj-daemon-set-bb6fd6fcb"
May 19 10:10:22.805: INFO: e2e-cf8wj-daemon-set-bb6fd6fcb has been updated
STEP: Generate another ControllerRevision by patching the Daemonset
STEP: Confirm that there are two ControllerRevisions
May 19 10:10:22.809: INFO: Requesting list of ControllerRevisions to confirm quantity
May 19 10:10:23.811: INFO: Requesting list of ControllerRevisions to confirm quantity
May 19 10:10:23.816: INFO: Found 2 ControllerRevisions
STEP: Removing a ControllerRevision via 'DeleteCollection' with labelSelector: "e2e-cf8wj-daemon-set-bb6fd6fcb=updated"
STEP: Confirm that there is only one ControllerRevision
May 19 10:10:23.821: INFO: Requesting list of ControllerRevisions to confirm quantity
May 19 10:10:23.823: INFO: Found 1 ControllerRevisions
```

# Verifying increase in coverage with APISnoop

## Listing endpoints hit by the new e2e test

This query shows the following endpoints are hit within a short period of running this e2e test

```sql-mode
select distinct  endpoint, right(useragent,51) AS useragent
from testing.audit_event
where endpoint ilike '%ControllerRevision%'
and release_date::BIGINT > round(((EXTRACT(EPOCH FROM NOW()))::numeric)*1000,0) - 60000
and useragent like 'e2e%should%'
order by endpoint
limit 10;
```

```example
                      endpoint                      |                      useragent
----------------------------------------------------+-----------------------------------------------------
 createAppsV1NamespacedControllerRevision           | should manage the lifecycle of a ControllerRevision
 deleteAppsV1CollectionNamespacedControllerRevision | should manage the lifecycle of a ControllerRevision
 deleteAppsV1NamespacedControllerRevision           | should manage the lifecycle of a ControllerRevision
 listAppsV1ControllerRevisionForAllNamespaces       | should manage the lifecycle of a ControllerRevision
 listAppsV1NamespacedControllerRevision             | should manage the lifecycle of a ControllerRevision
 patchAppsV1NamespacedControllerRevision            | should manage the lifecycle of a ControllerRevision
 readAppsV1NamespacedControllerRevision             | should manage the lifecycle of a ControllerRevision
 replaceAppsV1NamespacedControllerRevision          | should manage the lifecycle of a ControllerRevision
(8 rows)

```

# Final notes

If a test with these calls gets merged, **test coverage will go up by 7 points**

This test is also created with the goal of conformance promotion.

---

/sig testing
/sig architecture
/area conformance

### PR DESCRIPTION
**Pull request by riaankleinhans**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it:**
This PR adds a test to test the following untested endpoints:
- replaceAppsV1NamespacedControllerRevision
- readAppsV1NamespacedControllerRevision
- patchAppsV1NamespacedControllerRevision
- listAppsV1ControllerRevisionForAllNamespaces
- deleteAppsV1NamespacedControllerRevision 
- deleteAppsV1CollectionNamespacedControllerRevision 
- createAppsV1NamespacedControllerRevision 

**Which issue(s) this PR fixes:**
Fixes #110121

**Testgrid Link:** [Link](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=5&graph-metrics=test-duration-minutes&include-filter-by-regex=should%20manage%20the%20lifecycle%20of%20a%20ControllerRevision)

**Special notes for your reviewer:**
Adds +7 endpoint test coverage (good for conformance)

**Does this PR introduce a user-facing change?:**
```
NONE

```

**Release note:**
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:**
```
NONE

```

/sig testing
/sig architecture
/area conformance

### CODE REVIEW

File: test/e2e/apps/controller_revision.go
Comment (atiratree) [inline test/e2e/apps/controller_revision.go:137]: do we expect more than 1 revision?

File: test/e2e/apps/controller_revision.go
Comment (atiratree) [inline test/e2e/apps/controller_revision.go:254]: can we reuse this function from controller package without duplicating it here?

File: test/e2e/apps/controller_revision.go
Comment (atiratree) [inline test/e2e/apps/controller_revision.go:223]: could we also check the expected revision numbers?

File: test/e2e/apps/controller_revision.go
Comment (atiratree) [inline test/e2e/apps/controller_revision.go:133]: can you please comment that it is our intention to list in all namespaces

**Comment by heyste** [comment]
/test pull-kubernetes-e2e-kind-ipv6
unrelated flake

File: test/e2e/apps/controller_revision.go
Comment (heyste) [reply test/e2e/apps/controller_revision.go:135]: No, it was a leftover from some earlier testing. Fixed.

File: test/e2e/apps/controller_revision.go
Comment (heyste) [reply test/e2e/apps/controller_revision.go]: Fixed

File: test/e2e/apps/controller_revision.go
Comment (heyste) [reply test/e2e/apps/controller_revision.go:221]: I've included a check at the end of the test to confirm the expected revision number.

File: test/e2e/apps/controller_revision.go
Comment (heyste) [reply test/e2e/apps/controller_revision.go:131]: Comment added.

File: test/e2e/apps/controller_revision.go
Comment (atiratree) [reply test/e2e/apps/controller_revision.go:135]: in that case we can also remove the loop

Comment (heyste) [reply test/e2e/apps/controller_revision.go:135]: Fixed.

File: test/e2e/apps/controller_revision.go
Comment (atiratree) [inline test/e2e/apps/controller_revision.go:168]: I am getting the following error when I am running this localy. You need to change the deployment to create a unique hash of the daemon set to create the controller revision

```
  Jun 10 14:58:35.457: Failed to create ControllerRevision: controllerrevisions.apps "e2e-bd8m8-daemon-set-677bf9f4df" already exists
  Unexpected error:
      <*errors.StatusError | 0xc0039243c0>: {
          ErrStatus: {
              TypeMeta: {Kind: "", APIVersion: ""},
              ListMeta: {
                  SelfLink: "",
                  ResourceVersion: "",
                  Continue: "",
                  RemainingItemCount: nil,
              },
              Status: "Failure",
              Message: "controllerrevisions.apps \"e2e-bd8m8-daemon-set-677bf9f4df\" already exists",
              Reason: "AlreadyExists",
              Details: {
                  Name: "e2e-bd8m8-daemon-set-677bf9f4df",
                  Group: "apps",
                  Kind: "controllerrevisions",
                  UID: "",
                  Causes: nil,
                  RetryAfterSeconds: 0,
              },
              Code: 409,
          },
      }
      controllerrevisions.apps "e2e-bd8m8-daemon-set-677bf9f4df" already exists
  occurred
```

**Review by atiratree — COMMENTED** [review]
We are not getting the CI signal for a failure of this test since this test is running serially

**Review by soltysh — APPROVED** [review]
/approve
@atiratree's comment still stands, he'll be able to tag this once the problem is solved.

**Comment by aojea** [comment]
just FYI there are helpers already in the framework for working with daemonsets

https://github.com/kubernetes/kubernetes/blob/3375f5773b4371bc4975879521002b16ed8f0f56/test/e2e/framework/daemonset/fixtures.go#L65-L71

File: test/e2e/apps/controller_revision.go
Comment (aojea) [reply test/e2e/apps/controller_revision.go:53]: oh, so this is copied from this other test 

https://github.com/kubernetes/kubernetes/blob/0c93f75629ffb86cb7b2f2729b5124bfdff64c67/test/e2e/workload/daemon_set.go#L60-L65

@liggitt do you think this comment still applies?

**Comment by aojea** [comment]
/hold
until we can test it before merge
https://github.com/kubernetes/test-infra/pull/26556

File: test/e2e/apps/controller_revision.go
Comment (aojea) [inline test/e2e/apps/controller_revision.go:243]:
```suggestion
		if len(list.Items) != quantity {
			return false, nil
		}
```

Comment (aojea) [reply test/e2e/apps/controller_revision.go:243]: it's a nit but this way we don't have to do the mental exercise :smile:

File: test/e2e/apps/controller_revision.go
Comment (alculquicondor) [reply test/e2e/apps/controller_revision.go:106]: That can't be done for new tests. You need to proof test stability first. https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#promoting-tests-to-conformance

Comment (heyste) [reply test/e2e/apps/controller_revision.go:106]: When the current test has been merged and runs for 2 weeks without any flakes on test grid, then another PR will be created to promote the e2e test to conformance as linked in the above comment.

Comment (atiratree) [reply test/e2e/apps/controller_revision.go:106]: makes sense, thanks for the info

**Comment by heyste** [comment]
/test pull-kubernetes-conformance-kind-ga-only-parallel
unrelated flake timed out

**Comment by heyste** [comment]
/test pull-kubernetes-node-e2e-containerd
unrelated flake,  could not fetch resource

File: test/e2e/apps/controller_revision.go
Comment (heyste) [reply test/e2e/apps/controller_revision.go:168]: @atiratree Can we please confirm this error as my local testing doesn't create an error when I test it.

Also, I'm conscious of the time that this test needs to soak before it can be promoted. It would be awesome for all SIG-App endpoints to have conformance tests in this release.

File: test/e2e/apps/controller_revision.go
Comment (heyste) [reply test/e2e/apps/controller_revision.go]: @aojea Fixed. Is there anything blocking [https://github.com/kubernetes/test-infra/pull/26556](https://github.com/kubernetes/test-infra/pull/26556) at the moment? TIA

Comment (aojea) [reply test/e2e/apps/controller_revision.go]: ping @BenTheElder ^^^

File: test/e2e/apps/controller_revision.go
Comment (Riaankl) [reply test/e2e/apps/controller_revision.go:106]: @atiratree do you feel the /hold can be removed?

File: test/e2e/apps/controller_revision.go
Comment (atiratree) [reply test/e2e/apps/controller_revision.go:168]: I can still reproduce this with `./hack/local-up-cluster.sh`. It fails 100% times. 

Oddly, I am not getting any error with lastest kind.

Can you also please try with `local-up-cluster.sh` and debug the hashes and see what is the difference?

**Comment by Riaankl** [comment]
> I would prefere we test it in the CI before the merge, especially since we are getting different results on different environments ([#110122 (comment)](https://github.com/kubernetes/kubernetes/pull/110122#discussion_r911085399)). So it would be good if we could get another signal.

@atiratree my concern is that we are running out of time to get this merged, soaked and Promoted to Conformance before 1.25 test freeze. And we are keen to clean up the last SIG Apps endpoints with out Conformance test in this release. (2 weeks remaining on the time line)
@heyste have done some rigorous testing on the test and are confidant that it is solid. We have not had any of our previous test tested locally by a SIG, as they trust the test grid to show up any flakes.

Once this e2e test is merged we monitor it on the test grid for 2 weeks, if there is any flakes we investigate and update as needed. (As the test can not be promoted to conformance if there was as any flakes)

**Comment by aojea** [comment]
> /hold until we can test it before merge [kubernetes/test-infra#26556](https://github.com/kubernetes/test-infra/pull/26556)

/unhold

My apologies, I thought I was going.to be able to merge the optional job faster, doesn't makes any sense to keep this depending on the othr pr forever

**Comment by atiratree** [comment]
> my concern is that we are running out of time to get this merged, soaked and Promoted to Conformance before 1.25 test freeze. And we are keen to clean up the last SIG Apps endpoints with out Conformance test in this release. (2 weeks remaining on the time line)

let's at least wait for @heyste's response / result from running this test on his `local-up-cluster.sh`. 

also, I am sorry I am going away for 10 days, so I will pass this back to @soltysh

**Review by soltysh — APPROVED** [review]
Verified this manually, it's working as expected.
/lgtm

**Comment by Riaankl** [comment]
> @Riaankl apologies for the delays sweat_smile

Thanks a lot. Very exiting. 
This will close the last outstanding tests for Apps endpoints!!

We will monitor the test grid and Promote in two weeks time if no flakes appear.

**Comment by heyste** [comment]
/test pull-kubernetes-unit
unrelated flake: `kubelet/nodeshutdown`

**Comment by heyste** [comment]
/test pull-kubernetes-e2e-kind-ipv6
unrelated flake: `[sig-node] Probing container should mark readiness on pods to false and disable liveness probes while pod is in progress of terminating`

**Comment by heyste** [comment]
/test pull-kubernetes-e2e-gce-ubuntu-containerd
unrelated flakes: `e2e.go: TearDown` & `e2e.go: Timeout`

**Comment by Riaankl** [comment]
/unassign sig testing

(just testing the function)

**Pull request merged**
